### ISSUE TITLE
How to specify the file order?

### ISSUE THREAD

**Issue by tino** [benbria/karma-coffee-coverage#2]
Hi,

I have an existing project in which I use gulp to include my angular files in a certain order (all `*.module.coffee` files first), which works fine in the current setup with ibrik as follows:

```
module.exports = (config) ->
    config.set
        frameworks: ["jasmine"]

        files: [
            "../../static_root/build/js/vendor.js"
            "../../static_root/build/templates/templates.js"
            "coffee/*.coffee"
            "coffee/**/*.module.coffee"
            "coffee/**/*.app.coffee"
            "node_modules/angular-mocks/angular-mocks.js"
            "karma/django_constants.coffee"
            "coffee/**/*.spec.coffee"
            "coffee/**/!(*.spec).coffee"  # For code coverage
        ]

        # preprocess matching files before serving them to the browser
        preprocessors:
            "**/!(*.spec).coffee": ["coverage"]
            "**/*.spec.coffee": ["coffee"] # coverage converts the others with ibrik

        reporters: [
            "dots"
            "coverage"
        ]
        browsers: [
            "PhantomJS"
        ]
        coverageReporter:
            type: "html"
            dir: ".coverage/"
            # ibrik niet nodig, aangezien de coffee converter de files al convert naar js...
            instrumenters:
                ibrik: require('ibrik')
            instrumenter:
                '**/*.coffee': 'ibrik'
```

I can't get this converted to using coffee-coverage, which I would like, concerning the weird branch calculations with ibrik. When I just add:

```
    coffeeCoverage:
            framework:
                initAllSources: true,
                sourcesBasePath: 'coffee/'
                recursive: true
                dest: 'build/test/coverage-init.js',
                instrumentor: 'istanbul'
```
and add "build/test/coverage-init.js" to the files array, I get "reserved word var" a lot, ie, already processed coffee script files are attempted to process again.

When I comment out the preprocessors I get a parse error for every coffee file on the first line (ie, coffee files are read as js).

What to do?

**Comment by calvinwiebe** [comment]
Hmmm. Interesting. It is strange that the `preprocessors` would have any affect, as `karma-coffee-coverage` runs directly on the file system. 

I wonder if the `coverageReporter` is also building js files in the `coffee/` dir? 

I'll try and reproduce this, as it is a bit of a different use case than what we are using this plugin for. In the meantime, could you post the full karma config, and perhaps a tree structure of your project? 

Note: This plugin doesn't actually do much right now. That is, it doesn't even instrument the coffee files. All it is doing, right now, is generating this empty `coverage-init.js` so that [karma-coverage](https://github.com/karma-runner/karma-coverage) can take it and the actually instrumented `js` files specified in `files` and provide you a full report on your entire `sourceBasePath`, not just what you have tested. We use this plugin in conjunction with [browserify-coffee-coverage](https://www.npmjs.com/package/browserify-coffee-coverage). We build our browserify bundles, before running karma, and then just pass karma the instrumented bundles. 

What, I think, you want is to have your coffee files instrumented and compiled to js on the fly with `coffee-coverage`. Which should be easy enough. I'd just need to write a preprocessor for this repo that would be a drop in replacement for the `coffee` and `ibrik` preprocessors. Ja?

**Comment by tino** [comment]
Hi, the full conf that is working right now can be found here: https://gist.github.com/tino/851028f92b276fe211af

Directory structure:
```
.
├── Gulpfile.coffee
├── bower.json
├── bower_components
├── build
├── coffee
│   ├── base.module.coffee
│   ├── base.spec.coffee
....
│   ├── strings
│   │   ├── actions
│   │   │   ├── string.actions.coffee
│   │   │   ├── string.actions.spec.coffee
│   │   │   └── string.constants.coffee
│   │   ├── components
│   │   │   ├── buttons/
...
│   │   │   ├── string_new.coffee
│   │   │   └── string_new.spec.coffee
...
│   │   ├── strings.module.coffee
│   │   └── strings.module.spec.coffee
│   ├── styles.module.coffee
│   ├── systemen
│   └── urls.module.coffee
├── dist
├── images
├── karma
├── karma.conf.coffee
├── libs
├── node_modules
├── package.json
├── partials
└── scss
```

If it would be easy for you to write a preprocessor that would be great! I must say that even after having read the overview I am not sure I completely understand how this works :)

**Comment by calvinwiebe** [comment]
Hey @tino, I've written up a rudimentary preprocessor for this project. Are you familiar with `npm link`? If so, you can 
   * `git clone` this repo
   * `git checkout preprocessor`
   *  `npm link` it into your project. Then give it a go. 

[Here](https://gist.github.com/calvinwiebe/9fafa2499d2aefd94f34) is a `karma.conf.coffee` that I think should work for your project. Please look at the `preprocessors`, `coverageReporter` and `coffeeCoverage` props. Let me know if that does what you are looking for. 

As for your confusion after reading the overview, ya, I should make the readme a bit more clear. Essentially the `karma-coffee-coverage` _framework_ plugin that this project provides solves this problem:

You have a bunch of coffee source files. You have written tests for some of the files, but not all. You do not include all the coffee source files in your karma conf `files` list (common when browserifying your test files). This means the `karma-coverage` reporter has no idea about the other files in your project. So in its report you might have covered, say, 80% of `foo.coffee`, but you have not included `bar.coffee` in your tests. So your report shows 80%, but it should be 40% because of not testing `bar.coffee`. What to do? Then you use the _framework_ plugin in this project. It will walk through `sourcesBasePath` and generate a coverage object for _all_ of your src files (`foo.coffee` and `bar.coffee`) with empty counts and stick it into `coverage-init.js`. You then provide karma with this empty coverage object, and it combines it with what was actually tested against. Then your report will show you 40%, which is a much more accurate report.  

In your case, it seems like you are already including all of your source files in karma, so that part of the process isn't needed by you.

**Comment by tino** [comment]
Sweet, that works! :+1: Haha I expected the branch coverage statistics to go up using coffee-coverage instead of ibrik, but no... At least I can use the pragma's to make the output a bit more sensible ;)

Thanks for the fast response!

**Comment by calvinwiebe** [comment]
Excellent! I'll work on getting it into master and releasing a new version.

**Comment by calvinwiebe** [comment]
Released in `v1.1.0`.

**Comment by tino** [comment]
:+1:
